File: kotlin/6236.kt

fun main() {
    val (n, m) = readln().split(" ").map { it.toInt() }

    val arr = IntArray(n) { readln().toInt() }

    val sum = arr.sum().toLong()
    val maxValue = (arr.maxOrNull() ?: 0).toLong()

    val result = binarySearch(arr, sum, maxValue, m)
    print(result)
}

fun binarySearch(arr: IntArray, sum: Long, maxValue: Long, m: Int): Long {
    var lo = maxValue
    var hi = sum
    var result: Long = 0
    while (lo <= hi) {
        val mid = (lo + hi) / 2
        if (go(arr, mid, m)) {
            result = mid
            hi = mid - 1
        }
        else lo = mid + 1
    }
    return result
}

fun go(arr: IntArray, money: Long, m: Int): Boolean {
    var cnt = 1
    var tmp = money
    for (i in arr) {
        if (tmp - i < 0) {
            tmp = money
            cnt++
        }
        tmp -= i
    }
    return cnt <= m
}

File: CPP/baekjoon/6236.cpp
#include <iostream>
using namespace std;
typedef long long ll;
ll n, m, a[1000004], k, maxValue, lo, hi, sum;
bool go(int money) {
    int cnt = 1;
    int tmp = money;
    for (size_t i = 0; i < n; i++)
    {
        if (money - a[i] < 0) {
            money = tmp;
            cnt++;
        }
        money -= a[i];
    }
    return cnt <= m;
}
int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);
    cin >> n >> m;
    for (size_t i = 0; i < n; i++)
    {
        cin >> a[i];
        maxValue = max(maxValue, a[i]);
        sum += a[i];
    }
    lo = maxValue; hi = sum;
    while (lo <= hi)
    {
        ll mid = (lo + hi) / 2;
        if (go(mid)) {
            hi = mid - 1;
            k = mid;
        }
        else lo = mid + 1;
    }
    cout << k << '\n';
    
    return 0;
}
